### PR TITLE
Give the option to make `target_max_block_size` nullable

### LINKED ISSUE #47437: [data] ray.data.read_parquet not reading in one file per block
### What happened + What you expected to happen

I'd like to read in a set of files where each file corresponds to exactly one block in the dataset. I then want to use `map_batches` to process many files in parallel. However, when I read in parquet files using `ray.data.read_parquet` and set `override_num_blocks=NUM_FILES` I do not get each file processed as one block. Instead ray seems to be breaking my files into blocks that are 10,000 rows long, so if my file contains >10k rows it is split into multiple blocks, and then output in a garbled way. See code and output data below.

In the code example below I would expect all data within my single file to be pivoted together and to produce a single line of output. Instead, the first 10k rows of data appear on the first line, with additional rows from the input file appearing on additional lines. I need to find a way to pivot my large dataset so that all data for a given ID is output on a single line.

Note that if I apply `map_batches` to a Dataset I just defined (and did not read in) I get the behavior I expect: all data for a given ID is output on a single line.

Is there a way to enforce files be read in such that each file corresponds to exactly one block in the dataset? I see in [this issue](https://github.com/ray-project/ray/issues/33179#issuecomment-1463009674) that 1:1 mapping between files blocks was previously available by setting `parallelism=NUM_FILES`, but this no longer works. Nor does using the new `override_num_blocks`.

### Versions / Dependencies

Python 3.9.13
ray 2.30.0
MacOS Sonoma 14.6.1

### Reproduction script

```
import pandas as pd
import ray

# generate >10k lines of data
df = pd.DataFrame(data={"ID": "A", "values": range(0,10005), "dttm": pd.date_range("2024-01-01", periods=10005, freq="h")})
df = df.astype({"dttm": str})

# write out file
file_path = "/Desktop/ray_test_file.parquet"
df.to_parquet(file_path)

# read in file
ds = ray.data.read_parquet(
            file_path,
            override_num_blocks=1
        )

# define function
def _pivot_data(df: pd.DataFrame) -> pd.DataFrame:
    return df.pivot(index="ID", columns="dttm", values="values")

# transform data
ds_output = ds.map_batches(
            _pivot_data,  # type: ignore
            batch_size=None,
            batch_format="pandas"
        )
ds_output.to_pandas()
```
**Input dataset** looks like:

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ID</th>
      <th>values</th>
      <th>dttm</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>A</td>
      <td>0</td>
      <td>2024-01-01 00:00:00</td>
    </tr>
    <tr>
      <th>1</th>
      <td>A</td>
      <td>1</td>
      <td>2024-01-01 01:00:00</td>
    </tr>
    <tr>
      <th>2</th>
      <td>A</td>
      <td>2</td>
      <td>2024-01-01 02:00:00</td>
    </tr>
    <tr>
      <th>3</th>
      <td>A</td>
      <td>3</td>
      <td>2024-01-01 03:00:00</td>
    </tr>
    <tr>
      <th>4</th>
      <td>A</td>
      <td>4</td>
      <td>2024-01-01 04:00:00</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>10000</th>
      <td>A</td>
      <td>10000</td>
      <td>2025-02-20 16:00:00</td>
    </tr>
    <tr>
      <th>10001</th>
      <td>A</td>
      <td>10001</td>
      <td>2025-02-20 17:00:00</td>
    </tr>
    <tr>
      <th>10002</th>
      <td>A</td>
      <td>10002</td>
      <td>2025-02-20 18:00:00</td>
    </tr>
    <tr>
      <th>10003</th>
      <td>A</td>
      <td>10003</td>
      <td>2025-02-20 19:00:00</td>
    </tr>
    <tr>
      <th>10004</th>
      <td>A</td>
      <td>10004</td>
      <td>2025-02-20 20:00:00</td>
    </tr>
  </tbody>
</table>
<p>10005 rows × 3 columns</p>
</div>

**Returned dataset** looks like the following. It contains two rows, the first row containing the first 10k values and the second row containing the remaining 5 values. I expected/want all data for a given ID to be returned on a single row.

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th> dttm </th>
      <th>2024-01-01 00:00:00</th>
      <th>2024-01-01 01:00:00</th>
      <th>2024-01-01 02:00:00</th>
      <th>2024-01-01 03:00:00</th>
      <th>2024-01-01 04:00:00</th>
      <th>2024-01-01 05:00:00</th>
      <th>2024-01-01 06:00:00</th>
      <th>2024-01-01 07:00:00</th>
      <th>2024-01-01 08:00:00</th>
      <th>2024-01-01 09:00:00</th>
      <th>...</th>
      <th>2025-02-20 11:00:00</th>
      <th>2025-02-20 12:00:00</th>
      <th>2025-02-20 13:00:00</th>
      <th>2025-02-20 14:00:00</th>
      <th>2025-02-20 15:00:00</th>
      <th>2025-02-20 16:00:00</th>
      <th>2025-02-20 17:00:00</th>
      <th>2025-02-20 18:00:00</th>
      <th>2025-02-20 19:00:00</th>
      <th>2025-02-20 20:00:00</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>0.0</td>
      <td>1.0</td>
      <td>2.0</td>
      <td>3.0</td>
      <td>4.0</td>
      <td>5.0</td>
      <td>6.0</td>
      <td>7.0</td>
      <td>8.0</td>
      <td>9.0</td>
      <td>...</td>
      <td>9995.0</td>
      <td>9996.0</td>
      <td>9997.0</td>
      <td>9998.0</td>
      <td>9999.0</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>1</th>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>...</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>NaN</td>
      <td>10000.0</td>
      <td>10001.0</td>
      <td>10002.0</td>
      <td>10003.0</td>
      <td>10004.0</td>
    </tr>
  </tbody>
</table>
<p>2 rows × 10005 columns</p>
</div>

### Issue Severity

Medium: It is a significant difficulty but I can work around it.

### PR DESCRIPTION
**Pull request by goutamvenkat-anyscale**
## Why are these changes needed?

Ray Data enforces strict block sizing which may be counter-productive in cases where the partitions are already appropriately sized. 

## Related issue number

https://github.com/ray-project/ray/issues/47437 

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/data/context.py [hunk 1]
```diff
@@ -24,6 +25,8 @@
 _default_context: "Optional[DataContext]" = None
 _context_lock = threading.Lock()
 
+UNLIMITED_BLOCK_SIZE = sys.maxsize  # 9.22e18 - good enough for "∞"
```
Comment (alexeykudinkin) [inline python/ray/data/context.py]: That's not what we want -- we want to 

 - *Explicitly* set it to null
 - Make sure everywhere this config is used we handle it explicitly

File: python/ray/data/_internal/util.py [hunk 2]
```diff
@@ -245,10 +249,15 @@ def _autodetect_parallelism(
                 f"{ctx.target_min_block_size / (1024 * 1024)}MiB"
             )
         elif parallelism == min_safe_parallelism:
+            # Handle ``None`` (unlimited) gracefully in the log message.
+            if ctx.target_max_block_size is None:
+                display_val = "unlimited"
+            else:
+                display_val = f"{ctx.target_max_block_size / (1024 * 1024)}MiB"
```
Comment (srinathk10) [inline python/ray/data/_internal/util.py]: Good to hoist 1024 * 1024 up as a const and use. Also other consts in this func.

File: python/ray/data/_internal/datasource/parquet_datasource.py [hunk 3]
```diff
@@ -601,17 +601,23 @@ def compute_batch_size_rows(sample_info: _SampleInfo) -> int:
...
+                    ctx.target_max_block_size // 10
+                )
+                return max(
+                    1,
+                    min(
+                        PARQUET_READER_ROW_BATCH_SIZE,
+                        max_parquet_reader_row_batch_size_bytes
+                        // sample_info.actual_bytes_per_row,
+                    ),
+                )
```
Comment (srinathk10) [inline python/ray/data/_internal/datasource/parquet_datasource.py]: Good to hoist consts up.

File: python/ray/data/_internal/datasource/parquet_datasource.py [hunk 4]
```diff
@@ -601,17 +605,23 @@ def compute_batch_size_rows(sample_info: _SampleInfo) -> int:
...
-                    PARQUET_READER_ROW_BATCH_SIZE,
-                    max_parquet_reader_row_batch_size_bytes
-                    // sample_info.actual_bytes_per_row,
-                ),
-            )
+            ctx = DataContext.get_current()
+            if ctx.target_max_block_size is None:
+                # Use a large batch size to avoid splitting within a single block
+                # This should be large enough for most reasonable files
+                return LARGE_BATCH_SIZE
```
Comment (alexeykudinkin) [inline python/ray/data/_internal/datasource/parquet_datasource.py]: This should be null

File: python/ray/data/_internal/datasource/range_datasource.py [hunk 5]
```diff
@@ -47,6 +50,10 @@ def get_read_tasks(
         ctx = DataContext.get_current()
         if self._n == 0:
             target_rows_per_block = 0
```
Comment (alexeykudinkin) [inline python/ray/data/_internal/datasource/range_datasource.py]: Let's clean this up

File: python/ray/data/_internal/execution/operators/map_transformer.py [hunk 6]
```diff
@@ -102,6 +103,10 @@ def target_max_block_size(self):
         else:
             return self._output_block_size_option.target_max_block_size
 
+    @target_max_block_size.setter
+    def target_max_block_size(self, target_max_block_size: int):
+        self.set_target_max_block_size(target_max_block_size)
```
Comment (alexeykudinkin) [inline python/ray/data/_internal/execution/operators/map_transformer.py]: All entities should be immutable, let's avoid adding settes

File: python/ray/data/_internal/execution/operators/map_transformer.py [hunk 7]
```diff
@@ -220,6 +225,9 @@ def apply_transform(
         ctx: TaskContext,
     ) -> Iterable[Block]:
         """Apply the transform functions to the input blocks."""
+        if self.target_max_block_size is None:
+            self.set_target_max_block_size(sys.maxsize)
```
Comment (alexeykudinkin) [inline python/ray/data/_internal/execution/operators/map_transformer.py]: Let's avoid that (it doesn't solve the problems it masks them)

File: python/ray/data/_internal/logical/rules/set_read_parallelism.py [hunk 8]
```diff
@@ -21,6 +22,13 @@ def compute_additional_split_factor(
     target_max_block_size: int,
     cur_additional_split_factor: Optional[int] = None,
 ) -> Tuple[int, str, int, Optional[int]]:
+    """Returns parallelism to use and the min safe parallelism to avoid OOMs."""
+
+    # If the caller supplied ``None`` treat it as “no upper-limit”.
+    # Use ``sys.maxsize`` so that any arithmetic (/, //, *) works safely.
+    if target_max_block_size is None:
+        target_max_block_size = sys.maxsize
```
Comment (alexeykudinkin) [inline python/ray/data/_internal/logical/rules/set_read_parallelism.py]: See comment above, no maxsize fallbacks

File: python/ray/data/_internal/datasource/parquet_datasource.py [hunk 9]
```diff
@@ -482,7 +485,8 @@ def get_batch_iterable():
                 use_threads=use_threads,
                 columns=data_columns,
                 schema=schema,
-                batch_size=batch_size,
+                batch_size=batch_size
+                or ARROW_ALLOWABLE_MAX_BATCH_SIZE,  # By default, Pyarrow batch size is 131_072 and this parameter is not nullable. https://arrow.apache.org/docs/python/generated/pyarrow.dataset.Fragment.html#pyarrow.dataset.Fragment.to_batches
```
Comment (alexeykudinkin) [inline python/ray/data/_internal/datasource/parquet_datasource.py]: Let's avoid specifying one if it's none (use kwargs)

File: python/ray/data/context.py [hunk 10]
```diff
@@ -398,7 +398,7 @@ class DataContext:
             map tasks won't record memory stats.
     """
 
-    target_max_block_size: int = DEFAULT_TARGET_MAX_BLOCK_SIZE
+    target_max_block_size: Optional[int] = field(default=DEFAULT_TARGET_MAX_BLOCK_SIZE)
```
Comment (alexeykudinkin) [inline python/ray/data/context.py]:
```suggestion
    target_max_block_size: Optional[int] = DEFAULT_TARGET_MAX_BLOCK_SIZE
```

Comment (alexeykudinkin) [reply python/ray/data/context.py]: Let's keep it consistent w/ the rest (or fix them all separately)

File: python/ray/data/tests/test_parquet.py [hunk 11]
```diff
@@ -1640,6 +1640,84 @@ def test_parquet_row_group_size_002(ray_start_regular_shared, tmp_path):
     assert ds.fragments[0].num_row_groups == 10
 
 
+@pytest.mark.parametrize("override_num_blocks", [1, 2, 3])
+def test_max_block_size_none_respects_override_num_blocks(
```
Comment (alexeykudinkin) [inline python/ray/data/tests/test_parquet.py:1707]: Let's make sure we run test_read/write/map/group-by with disabled block-sizing.

You can do thta by adding a fixture (akin to configure_shuffle_method) to all of them

**Review by alexeykudinkin — COMMENTED** [review]
LGTM, let's expand test coverage and we'd be g2g

File: python/ray/data/_internal/datasource/parquet_datasource.py [hunk 12]
```diff
@@ -478,13 +478,16 @@ def read_fragments(
...
         def get_batch_iterable():
-            return fragment.to_batches(
-                use_threads=use_threads,
-                columns=data_columns,
-                schema=schema,
-                batch_size=batch_size,
+            kwargs = {
+                "use_threads": use_threads,
+                "columns": data_columns,
+                "schema": schema,
```
Comment (alexeykudinkin) [inline python/ray/data/_internal/datasource/parquet_datasource.py]: Let's only use kwargs for params that require it

File: python/ray/data/tests/test_parquet.py (hunk 11 above)
Comment (alexeykudinkin) [reply python/ray/data/tests/test_parquet.py:1707]: Let's add to all tests in these files

File: python/ray/data/_internal/datasource/range_datasource.py [hunk 13]
```diff
@@ -45,8 +48,10 @@ def get_read_tasks(
         # context if it was overridden. Set target max block size during
         # optimizer stage to fix this.
         ctx = DataContext.get_current()
-        if self._n == 0:
-            target_rows_per_block = 0
+        if ctx.target_max_block_size is None:
+            # If target_max_block_size is ``None``, treat it as unlimited and
+            # avoid further splitting.
+            target_rows_per_block = n  # whole block in one shot
```
Comment (iamjustinhsu) [inline python/ray/data/_internal/datasource/range_datasource.py:54]: curious, why would u want to set  target_rows_per_block = n in range dataset, if n is very large?

File: python/ray/data/_internal/datasource/parquet_datasource.py [hunk 14]
```diff
@@ -478,13 +478,16 @@ def read_fragments(
...
                 **to_batches_kwargs,
-            )
+            }
+            if batch_size is not None:
+                kwargs["batch_size"] = batch_size
```
Comment (iamjustinhsu) [inline python/ray/data/_internal/datasource/parquet_datasource.py]: question: is the default batch size https://arrow.apache.org/docs/python/generated/pyarrow.dataset.ParquetFileFragment.html#pyarrow.dataset.ParquetFileFragment.to_batches a good estimation?

File: python/ray/data/_internal/datasource/range_datasource.py (hunk 13 above)
Comment (goutamvenkat-anyscale) [reply python/ray/data/_internal/datasource/range_datasource.py:54]: Same principle as other data sources. If the user wants to explicitly override target_max_block_size to be unlimited we give them the option to use the full range as 1 block

File: python/ray/data/context.py [hunk 15]
```diff
@@ -398,7 +398,7 @@ class DataContext:
             map tasks won't record memory stats.
     """
 
-    target_max_block_size: int = DEFAULT_TARGET_MAX_BLOCK_SIZE
+    target_max_block_size: Optional[int] = DEFAULT_TARGET_MAX_BLOCK_SIZE
```
Comment (raulchen) [inline python/ray/data/context.py:402]: can you also update the comment?

**Comment by raulchen** [comment]
discussed with @goutamvenkat-anyscale offline. 
While this PR per se makes sense (allow target_max_block_size to be infinite), it doesn't address https://github.com/ray-project/ray/issues/47437
Because when users set `override_num_blocks` or set num_partitions for an aggregation op, we should only disable splitting for the read/agg op. Users don't want to disable splitting for the rest ops in the pipeline. 
We need another PR to handle that.

File: python/ray/data/_internal/util.py [hunk 16]
```diff
@@ -201,9 +201,14 @@ def _autodetect_parallelism(
...
     min_safe_parallelism = 1
     max_reasonable_parallelism = sys.maxsize
+
     if mem_size is None and datasource_or_legacy_reader:
         mem_size = datasource_or_legacy_reader.estimate_inmemory_data_size()
-    if mem_size is not None and not np.isnan(mem_size):
+    if (
+        mem_size is not None
+        and not np.isnan(mem_size)
+        and target_max_block_size is not None
```
Comment (raulchen) [inline python/ray/data/_internal/util.py:210]: also update the parameter type

File: python/ray/data/_internal/logical/rules/set_read_parallelism.py [hunk 17]
```diff
@@ -34,7 +36,13 @@ def compute_additional_split_factor(
         logger.debug(
             f"Expected in-memory size {mem_size}," f" block size {expected_block_size}"
         )
-        size_based_splits = round(max(1, expected_block_size / target_max_block_size))
+        if target_max_block_size is None:
```
Comment (raulchen) [inline python/ray/data/_internal/logical/rules/set_read_parallelism.py:39]: same here - paramter type

File: python/ray/data/_internal/execution/operators/map_transformer.py [hunk 18]
```diff
@@ -90,7 +90,6 @@ def output_block_size_option(self):
         return self._output_block_size_option
 
     def set_target_max_block_size(self, target_max_block_size: int):
```
Comment (raulchen) [inline python/ray/data/_internal/execution/operators/map_transformer.py]: parameter type

File: python/ray/data/_internal/execution/operators/hash_shuffle.py [hunk 19]
```diff
@@ -809,15 +809,17 @@ def _get_default_aggregator_ray_remote_args(
...
-            num_aggregators=num_aggregators,
-            num_partitions=num_partitions,
-            # NOTE: If no partition size hint is provided we simply assume target
-            #       max block size specified as the best partition size estimate
-            partition_byte_size_estimate=(
-                partition_size_hint or self.data_context.target_max_block_size
-            ),
-        )
+        aggregator_total_memory_required = 0
+        if self.data_context.target_max_block_size is not None:
```
Comment (raulchen) [inline python/ray/data/_internal/execution/operators/hash_shuffle.py]: for hash shuffle, we should never use target_max_block_size because `num_partitions` is already specified

File: python/ray/data/_internal/execution/interfaces/physical_operator.py [hunk 20]
```diff
@@ -317,7 +317,7 @@ def target_max_block_size(self) -> Optional[int]:
             return self._output_block_size_option.target_max_block_size
 
     @property
-    def actual_target_max_block_size(self) -> int:
+    def actual_target_max_block_size(self) -> Optional[int]:
```
Comment (raulchen) [inline python/ray/data/_internal/execution/interfaces/physical_operator.py:320]: please also update the docstring

File: python/ray/data/tests/test_csv.py [hunk 21]
```diff
@@ -67,7 +67,9 @@ def test_csv_read_partitioning(ray_start_regular_shared, tmp_path):
         ),
     ],
 )
-def test_csv_read(ray_start_regular_shared, fs, data_path, endpoint_url):
+def test_csv_read(
+    ray_start_regular_shared, fs, data_path, endpoint_url, target_max_block_size_none
```
Comment (raulchen) [inline python/ray/data/tests/test_csv.py]: why do we want to set it to None for these csv tests?

File: python/ray/data/tests/test_groupby_e2e.py [hunk 22]
```diff
@@ -48,7 +48,9 @@ def _sort_series_of_lists_elements(s: pd.Series):
 
 
 def test_grouped_dataset_repr(
-    ray_start_regular_shared_2_cpus, disable_fallback_to_object_extension
+    ray_start_regular_shared_2_cpus,
+    disable_fallback_to_object_extension,
+    target_max_block_size_none,
```
Comment (raulchen) [inline python/ray/data/tests/test_groupby_e2e.py]: same here, why none?

File: python/ray/data/tests/test_parquet.py [hunk 23]
```diff
@@ -1640,9 +1697,72 @@ def test_parquet_row_group_size_002(ray_start_regular_shared, tmp_path):
...
+    ray_start_regular_shared, tmp_path, override_num_blocks, target_max_block_size_none
+):
+    """
+    When `DataContext.target_max_block_size` is explicitly set to ``None``,
```
Comment (raulchen) [inline python/ray/data/tests/test_parquet.py:1714]: per my last comment, override_num_blocks should always be respected even when `target_max_block_size` isn't set to None. 
We should remove the fixture in the follow-up PR.
Also, I'd like to add another test case that checks `target_max_block_size` is disabled globally when set to None.

**Review by raulchen — COMMENTED** [review]
Mostly LGTM, except some small comments.

File: python/ray/data/_internal/execution/operators/hash_shuffle.py (hunk 19 above)
Comment (goutamvenkat-anyscale) [reply python/ray/data/_internal/execution/operators/hash_shuffle.py]: This is the same problem as `override_num_blocks`. I can address this is in a separate PR if that's alright

File: python/ray/data/tests/test_csv.py (hunk 21 above)
Comment (goutamvenkat-anyscale) [reply python/ray/data/tests/test_csv.py]: This was based of an offline conversation. Essentially we wanted to add this test fixture to csv, json, groupby, map, parquet operations to ensure that infinite block sizing doesn't break any existing behavior. Of course the concern is the additional cost for CI runs. If you prefer I can reduce the number of tests that employ this fixture.

File: python/ray/data/tests/test_groupby_e2e.py (hunk 22 above)
Comment (goutamvenkat-anyscale) [reply python/ray/data/tests/test_groupby_e2e.py]: Commented here: https://github.com/ray-project/ray/pull/54450#discussion_r2220716401

File: python/ray/data/tests/test_parquet.py (hunk 23 above)
Comment (goutamvenkat-anyscale) [reply python/ray/data/tests/test_parquet.py:1714]: Added a test in `test_block_sizing.py` , test_target_max_block_size_none_disables_splitting_globally

File: python/ray/data/tests/test_csv.py (hunk 21 above)
Comment (raulchen) [reply python/ray/data/tests/test_csv.py]: The fixture seems to always set the target block size to None. 
I guess we want to parameterize this config - testing both default and None, right?

File: python/ray/data/_internal/execution/operators/hash_shuffle.py (hunk 19 above)
Comment (raulchen) [reply python/ray/data/_internal/execution/operators/hash_shuffle.py]: Fine to address that later. 
For this PR, we should also handle this when partition_size_hint is not None
```
partition_byte_size_estimate=(
    partition_size_hint or self.data_context.target_max_block_size
)
if partition_byte_size_estimate is not None: 
   ...
```

File: python/ray/data/tests/test_block_sizing.py [hunk 24]
```diff
@@ -252,6 +252,34 @@ def test_shuffle(shutdown_only, restore_data_context, shuffle_op):
...
+    ctx = DataContext.get_current()
+    ctx.target_max_block_size = 1_000_000  # 1MB - much smaller than data
+
+    ds_with_limit = ray.data.range(large_data_size, override_num_blocks=1).materialize()
+    blocks_with_limit = ds_with_limit._plan.initial_num_blocks()
+
+    # Now test with target_max_block_size = None (should not split)
+    ctx.target_max_block_size = None  # Disable block size limit
+
+    ds_unlimited = ray.data.range(large_data_size, override_num_blocks=1).materialize()
```
Comment (raulchen) [inline python/ray/data/tests/test_block_sizing.py]: can you also add a no-op map before the materialize?

File: python/ray/data/tests/test_parquet.py (hunk 23 above)
Comment (raulchen) [reply python/ray/data/tests/test_parquet.py:1714]: SG. please add a TODO here.

**Review by raulchen — APPROVED** [review]

File: python/ray/data/_internal/execution/operators/hash_shuffle.py (hunk 19 above)
Comment (goutamvenkat-anyscale) [reply python/ray/data/_internal/execution/operators/hash_shuffle.py]: In this block `partition_byte_size_estimate` is guaranteed to be non-None because `self.data_context.target_max_block_size` is not `None` when it's executed in this case.

File: python/ray/data/tests/test_parquet.py (hunk 23 above)
Comment (goutamvenkat-anyscale) [reply python/ray/data/tests/test_parquet.py:1714]: Added TODO

File: python/ray/data/tests/test_csv.py (hunk 21 above)
Comment (goutamvenkat-anyscale) [reply python/ray/data/tests/test_csv.py]: So instead of this pytest fixture, I can parameterize select tests to use `@pytest.mark.parametrize("target_max_block_size", [None, DEFAULT_TARGET_MAX_BLOCK_SIZE])`. Does that work?

File: python/ray/data/_internal/execution/operators/hash_shuffle.py (hunk 19 above)
Comment (goutamvenkat-anyscale) [reply python/ray/data/_internal/execution/operators/hash_shuffle.py]: addressed by handling `partition_byte_size_estimate` case when it's not None.

File: python/ray/data/tests/conftest.py [hunk 25]
```diff
@@ -360,6 +360,16 @@ def target_max_block_size(request):
     ctx.target_max_block_size = original
 
 
+@pytest.fixture(params=[None, DEFAULT_TARGET_MAX_BLOCK_SIZE])
+def target_max_block_size_none():
+    """Fixture that sets target_max_block_size to None and resets after test finishes."""
+    ctx = ray.data.context.DataContext.get_current()
+    original = ctx.target_max_block_size
+    ctx.target_max_block_size = None
```
Comment (raulchen) [inline python/ray/data/tests/conftest.py]:
```suggestion
@pytest.fixture(params=[None, DEFAULT_TARGET_MAX_BLOCK_SIZE])
def target_max_block_size_none(request):
    """Fixture that sets target_max_block_size to None and resets after test finishes."""
    ctx = ray.data.context.DataContext.get_current()
    original = ctx.target_max_block_size
    ctx.target_max_block_size = request.param
```

Comment (raulchen) [reply python/ray/data/tests/conftest.py]: and can you also update the name and docstring?

Comment (goutamvenkat-anyscale) [reply python/ray/data/tests/conftest.py]: Yes renamed to `target_max_block_size_infinite_or_default`
